init/tmpfiles.py

```diff
@@ -0,0 +1,606 @@
+            with open(path, 'w') as f:
+                f.write(self.arg if self.arg is not None else '')
+        self._chown_and_chmod(path)
+
+
+class FileCreateOrTruncate(Action):
+    """f+ - create or truncate file, with optional content"""
+    def apply_one(self, path):
+        with open(path, 'w') as f:
+            f.write(self.arg if self.arg is not None else '')
+        self._chown_and_chmod(path)
+
+
+class FileWrite(GlobAction):
+    """w - write to file"""
+    # TODO interpret C-style blackslashes in argument.  Also for
+    # other actions (f, f+, w+, ...)
+    def apply_one(self, path):
+        with open(path, 'w') as f:
+            f.write(self.arg if self.arg is not None else '')
+        self._chown_and_chmod(path)
+
+
+class FileAppend(GlobAction):
+    """w+ - append to file"""
+    def apply_one(self, path):
+        with open(path, 'a') as f:
+            f.write(self.arg if self.arg is not None else '')
+        self._chown_and_chmod(path)
+
+
+class DirCreateAndCleanup(Action):
+    """d - create and cleanup directory"""
+    def apply_one(self, path):
+        if not os.path.lexists(path):
+            os.makedirs(path)
+            self._chown_and_chmod(path)
+
+
+class DirCreateAndRemove(DirCreateAndCleanup):
+    """D - create and remove directory"""
+    # No additional behaviour, until --remove gets implemented
+
+
+class DirCleanup(GlobAction):
+    """e - create and remove directory"""
+    def apply_one(self, path):
+        if os.path.isdir(path):
+            self._chown_and_chmod(path)
+
+
+class SubvolumeCreate_v(DirCreateAndCleanup):
+    """v - create subvolume or directory"""
+    # ignore subvolume behaviour
+
+
+class SubvolumeCreate_q(DirCreateAndCleanup):
+    """q - create subvolume or directory"""
+    # ignore subvolume behaviour
+
+
+class SubvolumeCreate_Q(DirCreateAndCleanup):
+    """Q - create subvolume or directory"""
+    # ignore subvolume behaviour
+
+
+# TODO p, p+
+
+
+class SymlinkCreate(Action):
+    """L - create symlink"""
+    def apply_one(self, path):
+        if not os.path.exists(path):
+            if self.arg is None:
+                # TODO link to /usr/share/factory/FILE
+                # (see tmpfiles.d(5) for details)
+                raise RuntimeError("symlink target not specified")
+            os.symlink(self.arg, path)
+
+
+class SymlinkRecreate(SymlinkCreate):
+    """L+ - [re]create symlink"""
+    def apply_one(self, path):
+        if os.path.exists(path):
+            if not os.path.islink(path) and os.path.isdir(path):
+                shutil.rmtree(path)
+            else:
+                os.unlink(path)
+        super().apply(path)
+
+
+class CreateCharDev(Action):
+    """c - create character device node"""
+
+    arg_function = staticmethod(_parse_major_minor)
+
+    def apply_one(self, path):
+        major, minor = self.args
+        if not os.path.lexists():
+            subprocess.run(['mknod', path, c, major, minor])
```

Comment:
``suprocess.run`` does not raise an exception on error. It only signals error in return code. Use ``check_call`` to run a command and check its return code:

```
>>> subprocess.run(["false"])
CompletedProcess(args=['false'], returncode=1)
>>> subprocess.check_call(["false"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['false']' returned non-zero exit status 1.
```

```suggestion
            subprocess.check_call(['mknod', path, c, major, minor])
```

Reply:
I factored out the subprocess invocation to a new top-level subroutine `_run_process`.  It emits a message when the call fails due to executable not found, or upon non-zero exit status of subprocess.